library/src/main/scala/scala/scalajs/js/Thenable.scala

```diff
@@ -80,4 +84,23 @@ object Thenable {
     implicit def thenable2future[A](p: Thenable[A]): Future[A] =
       p.toFuture
   }
+
+  sealed trait Returning[A, B]
+
+  trait Returning1 {
+    implicit def default[A]: Returning[A, A] = new Returning[A, A] {}
+  }
+
+  object Returning extends Returning1 {
+
+    implicit def flattened[A, T[X] <: Thenable[X]]: Returning[T[A], A] =
+      new Returning[T[A], A] {}
```

Comment:
Your implementation of `mapP` does not compile: you have to provide an evidence that resolving a `B` effectively returns a `js.Promise[B]`. You can do that by adding an implicit parameter of type `Thenable.Returning[B, C]`, and it will fix the return type of `mapP` to `js.Promise[C]`, not `js.Promise[B]`.

In other words, and that’s the point of this PR, you can not implement `Functor[js.Promise]` because `js.Promise` breaks parametricity.